src/util/object_forwarding.rs

```diff
@@ -81,8 +81,6 @@ pub fn forward_object<VM: VMBinding>(
     copy_context: &mut GCWorkerCopyContext<VM>,
 ) -> ObjectReference {
     let new_object = VM::VMObjectModel::copy(object, semantics, copy_context);
-    #[cfg(feature = "vo_bit")]
-    crate::util::metadata::vo_bit::set_vo_bit::<VM>(new_object);
```

Comment:
Fully evacuating policies, such as SemiSpace and GenCopy, only needs to set VO bits for to-space objects during GC, and bulk clear VO bits for the from space (and the nursery) after GC.  The reconstruction part is similar to the `ClearAndReconstruct` strategy, but also makes VO bits available during tracing as it clears the VO bits after tracing rather than before.

I'll have a closer look at MarkCompact.  It should be able to clear and reconstruct the VO bits either during the CalculateForwarding stage (when computing the destination of each object) or the second transitive closure.  It should be easier to do it in CalcualteForwarding.  The marking phase of MarkCompact is non-moving, so the VO bits remain valid during that phase, which can help us validate the object graph using VO bits and dump heap by tracing during that phase.  However, MarkCompact does not support conservative GC as it does not support object pinning, so we don't need to worry about conservative roots or fields during the second transitive closure.

Reply:
The current implementation of MarkCompact unconditionally uses the VO bits regardless whether the Cargo feature "vo_bit" is enabled or not (it should use a local metadata if the VM binding doesn't demand global VO bits).  The VO bit is set in `post_alloc`.  During compaction (`MarkCompactSpace::compact`), it performs linear scanning using VO bits.  For each address with VO bits set (where an object, live or dead, exists), it first clears the VO bit, and then if the object is alive (forwarded), it will set the VO bit on its new location.  By doing this, it ensures at the end of compaction, VO bits are set for all live objects, but cleared for all dead objects.

If I enable sanity GC and VO bits, and then run OpenJDK with MarkCompact, I see all live objects have VO bits set.

According to this algorithm, the VO bits are available during both marking and forwarding, because objects are not moved and the VO bits are not modified during those two phases.